src/object/srv_ec_aggregate.c

```diff
@@ -334,6 +399,155 @@ agg_sgl_fini(d_sg_list_t *sgl)
+			if (!isset(tbit_map, i)) {
+				setbit(tbit_map, i);
+				cell_cnt++;
+			}
+		}
+	}
+
+	return cell_cnt;
+}
+
+/* Initializes the object handle of for the object represented by the entry.
+ * No way to do this until pool handle uuid and container handle uuid are
+ * initialized and share to other servers at higher(pool/container) layer.
+ */
+static int
+agg_get_obj_handle(struct ec_agg_entry *entry)
+{
+	struct daos_obj_layout	*layout;
+	struct ec_agg_param	*agg_param;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	int			 i, j, rc = 0;
+	bool			 opened = false;
+
+	agg_param = container_of(entry, struct ec_agg_param, ap_agg_entry);
+	if (daos_handle_is_inval(entry->ae_obj_hdl)) {
+		rc = dsc_obj_open(agg_param->ap_pool_info.api_cont_hdl,
+				  entry->ae_oid.id_pub, DAOS_OO_RW,
+				  &entry->ae_obj_hdl);
+		if (!rc)
+			opened = true;
+	}
+	if (opened) {
+		d_rank_t myrank;
+
+		crt_group_rank(NULL, &myrank);
+		dc_obj_layout_get(entry->ae_obj_hdl, &layout);
+		for (i = 0; i < layout->ol_nr; i++)
+			for (j = 0; j < layout->ol_shards[i]->os_replica_nr;
+			     j++) {
+				int p = 0;
+
+				if (j >= k && layout->ol_shards[i]->
+				    os_shard_loc[j].sd_rank != myrank) {
+					entry->ae_peer_pshards[p].sd_rank
+						= layout->ol_shards[i]->
+						os_shard_loc[j].sd_rank;
+					entry->ae_peer_pshards[p++].sd_tgt_idx
+						= layout->ol_shards[i]->
+						os_shard_loc[j].sd_tgt_idx;
+				}
+			}
+		daos_obj_layout_free(layout);
+	}
+	return rc;
+}
+
+/* Fetches the old data for the cells in the stripe undergoing a partial parity
+ * update, or a parity recalculation. For update, the bit_map indicates the
+ * cells that are present as replicas. In this case the parity epoch is used
+ * for the fetch. For recalc, the bit_map indicates the cells that are not fully
+ * populated as replicas. In this case, the highest replica epoch is used.
+ */
+static int
+agg_fetch_odata_cells(struct ec_agg_entry *entry, uint8_t *bit_map,
+		      unsigned int cell_cnt, bool is_recalc)
+{
+	daos_iod_t		 iod = { 0 };
+	daos_epoch_t		 epoch = { 0 };	/* epoch used for data fetch */
+	daos_recx_t		*recxs = NULL;
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	unsigned int		 i, j;
+	int			 rc = 0;
+
+	D_ALLOC_ARRAY(recxs, cell_cnt);
+	if (recxs == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+	for (i = 0, j = 0; i < k; i++)
+		if (isset(bit_map, i)) {
+			recxs[j].rx_idx =
+			  entry->ae_cur_stripe.as_stripenum * k * len + i * len;
+			recxs[j++].rx_nr = len;
+		}
+	D_ASSERT(j == cell_cnt);
+
+	iod.iod_name	= entry->ae_akey;
+	iod.iod_type	= DAOS_IOD_ARRAY;
+	iod.iod_size	= entry->ae_rsize;
+	iod.iod_nr	= cell_cnt;
+	iod.iod_recxs	= recxs;
+	entry->ae_sgl.sg_nr = 1;
+	entry->ae_sgl.sg_iovs[AGG_IOV_ODATA].iov_len = cell_cnt * len *
+								entry->ae_rsize;
+	(entry->ae_sgl.sg_iovs)++;
+
+	rc = agg_get_obj_handle(entry);
+	if (rc) {
+		D_ERROR("Failed to open object: "DF_RC"\n", DP_RC(rc));
+		goto out;
```

Comment:
revert sg_iovs back?

Reply:
Yep, good catch.  Thanks.

Reply:
it seems to be -- ?

Reply:
Yeah, I copied the line, forgot to change the operator...it's fixed now.